src/DefinitionResolver.php

```diff
@@ -479,8 +483,10 @@ public function resolveExpressionNodeToType(Node\Expr $expr): Type
                 if ($expr instanceof Node\Expr\MethodCall) {
                     $fqn .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
+                if ($def = $this->index->getDefinition($fqn)) {
+                    if ($def->type instanceof Types\This || $def->type instanceof Types\Self_) {
+                        return $this->resolveExpressionNodeToType($expr->var);
+                    }
```

Comment:
@felixfbecker should I remove the Self_ check? Or do you need more time to think about it?